python/tvm/topi/cuda/nms.py

```diff
@@ -54,64 +54,66 @@ def atomic_add(x, y):
+def rearrange_indices_out_ir(data, output, valid_box_count):
     """Hybrid routine to rearrange nms output to
     move all valid entries to top.
 
     Parameters
     ----------
     data : tvm.te.Tensor or numpy NDArray
+        NMS output. 3-D tensor with shape
+        [batch_size, num_anchors, 6] or
+        [batch_size, num_anchors, 5], or 2-D
         tensor with shape [batch_size, num_anchors].
 
+    one: tvm.tir.const
+        Constant one with the same dtype as data.
+
+    batch_size: tvm.tir.IntImm or tvm.tir.Var
+        Batch size. We need to pass it in since hybrid script doesn't support
+        binding variable to symbolic dim.
+
+    num_anchors: tvm.tir.IntImm or tvm.tir.Var
+        Number of anchors.
 
     Returns
     -------
-    stmt : Stmt
-        The result IR statement.
+    output : tvm.te.Tensor or numpy NDArray
+        2-D tensor with shape [batch_size, num_anchors].
+
+    valid_box_count : tvm.te.Tensor or numpy NDArray
+        Tensor with shape [batch_size, 1], indicates
+        the valid number of boxes.
     """
     batch_size = data.shape[0]
     num_anchors = data.shape[1]
 
     ib = tvm.tir.ir_builder.create()
+
     data = ib.buffer_ptr(data)
-    out = ib.buffer_ptr(out)
     valid_box_count = ib.buffer_ptr(valid_box_count)
-
-    one_count = tvm.tir.const(1, dtype="int32")
-    atomic_add_return = ib.allocate(
-        valid_box_count.dtype, (batch_size,), name="atomic_add_return", scope="local"
-    )
-
-    max_threads = int(tvm.target.Target.current(allow_none=False).max_num_threads)
-    nthread_tx = max_threads
-    tx = te.thread_axis("threadIdx.x")
-    ib.scope_attr(tx, "thread_extent", nthread_tx)
-    len_inner_for = (batch_size * num_anchors) // nthread_tx + 2
-
-    idxd = tvm.tir.indexdiv
-    idxm = tvm.tir.indexmod
-
-    with ib.for_range(0, len_inner_for, name="i") as i:
-        idx = tx * len_inner_for + i
-        batch_idx = idxd(idx, num_anchors)
-        with ib.if_scope(idx < batch_size):
-            valid_box_count[idx] = 0
-        with ib.if_scope(idx < batch_size * num_anchors):
-            with ib.if_scope(data[idx] >= 0):
-                atomic_add_return[batch_idx] = atomic_add(
-                    tvm.tir.call_intrin("handle", "tir.address_of", valid_box_count[batch_idx]),
-                    one_count,
-                )
-                out[batch_idx * num_anchors + atomic_add_return[batch_idx]] = data[idx]
-            with ib.if_scope(tvm.tir.any(data[idx] > num_anchors, data[idx] < -num_anchors)):
-                atomic_add_return[batch_idx] = atomic_add(
-                    tvm.tir.call_intrin("handle", "tir.address_of", valid_box_count[batch_idx]),
-                    one_count,
-                )
-                out[batch_idx * num_anchors + atomic_add_return[batch_idx]] = 0
-
-            with ib.if_scope(idxm(idx, num_anchors) >= valid_box_count[batch_idx]):
-                out[idx] = -1
```

Comment:
This implementation of rearrange_indices_out_ir returns an undersized tensor in some case, I think the threading isn't quite right, but i haven't been able to fix.